Refactor: Collect render

Tilix4 opened the pull request:
## Changelog Description
Refactored render collection code structure to allow building custom render collection over existing code.
The main goal is to drive which frames are rendered from the instance data instead of global `context` data.

Created a `CollectHarmonyRender` which is the base class for collecting Harmony render.

Also cleaned useless complexity like `HarmonyRenderInstance`. Also refactored JS API to a shared `getRenderNodeSettings` function instead of these strange api functions based on class name, which are the main limitation for inheriting from classes.

## Additional review information
**I had the `collect_farm_render.py` refactored but I don't have the possibility to test it, so please do it!**

Also refactored code on core, which ends up being cleaner, but not required: https://github.com/ynput/ayon-core/pull/1282

## Testing notes:
1. Publish render on Harmony

Tilix4 committed:
Refactor: Collect render

Tilix4 committed:
linting

Tilix4 committed:
linting

Tilix4 committed:
linting

kalisp commented:
That `useless complexity like HarmonyRenderInstance` was there because of Deadline submission where you need to pass format of rendered files. (If I remember correctly.)

BigRoy commented on client/ayon_harmony/plugins/publish/extract_render.py:
```diff
@@ -116,7 +108,22 @@ def process(self, instance):
+        self.log.debug(f"collections: {collections}")
+        if len(collections) > 1:
+            for col in reversed(collections):
+                if len(list(col)) > 1:
```
```suggestion
                if len(col.indexes) > 1:
```

BigRoy replied:
But is there really no better way to _know_ what files were actually rendered? :')

Tilix4 replied:
I don't know, I copied this from the previous version of the code. I think we could either parse the process output, or build a list of expected files with the parameters passed to the `Popen` process. But this is an optimization not directly linked to this PR.

Tilix4 committed:
Optimize length check

Co-authored-by: Roy Nieterau <roy_nieterau@hotmail.com>

Tilix4 commented:
> That `useless complexity like HarmonyRenderInstance` was there because of Deadline submission where you need to pass format of rendered files. (If I remember correctly.)

I believe you, but as far as there is no any comment about that, I couldn't know and only consider it as superfluous. If you point me out the related deadline code I'll consider a design taking it into consideration!

Tilix4 committed:
merge get_instances code to parent class

Tilix4 committed:
better explanation for render collection selection

BigRoy commented on client/ayon_harmony/plugins/publish/collect_render_instances.py:
```diff
@@ -0,0 +1,18 @@
+from ayon_harmony.api.plugins import publish
+
+
+class CollectRender(publish.CollectHarmonyRender):
```
It is in `api/` though? But yeah, I had the same confusion at first. It's just in a subfolder of `api/`. It's **not** next to the publish plug-ins.

iLLiCiTiT replied:
> It is in api/ though? 

Oh, the naming is confugis as hell. Ok, please rename `CollectHarmonyRender` to `CollectRenderBase` and rename `collect_harmony_render.py` to just `render.py`. Also put both collectors to one file, and confirm that `CollectRender` should run for both local and farm instances (Because I don't understand why it runs twice?).

Tilix4 replied:
> Is it ok that this will be triggered for both local and farm instances?

Could you please elaborate? I have no functional farm so I cannot test, I rely on you for this.

iLLiCiTiT replied:
The plugin has filter `families = ["render"]`. Both farm and local instances have family `"render"`.

Tilix4 replied:
Ok got it, you're right it shouldn't run twice for farm.

Tilix4 committed:
both collect render and farm in same file

iLLiCiTiT commented on client/ayon_harmony/plugins/publish/collect_render.py:
```diff
@@ -1,9 +1,23 @@
+    label = "Collect Render Instances"
+    order = pyblish.api.CollectorOrder + 0.01
+    hosts = ["harmony"]
+    families = ["render"]
```
```suggestion
    families = ["render.local"]
```

Tilix4 committed:
Collect render only for local

Co-authored-by: Jakub Trllo <43494761+iLLiCiTiT@users.noreply.github.com>

Tilix4 committed:
renaming

Tilix4 committed:
Merge branch 'refactor-collect_render' of https://github.com/Tilix4/ayon-harmony into refactor-collect_render

iLLiCiTiT commented on client/ayon_harmony/plugins/publish/collect_render.py:
```diff
@@ -0,0 +1,45 @@
+    label = "Collect Farm Render"
+    families = ["render.farm"]
+
+    def add_additional_data(self, instance):
```
It looks like this is the only difference in `CollectFarmRender` and `CollectRender`?

If that's true then we don't need 2 plugins. Just move `CollectRenderBase` here, rename it to `CollectHarmonyRenderInstances`, add `families =["render"]` filter and add method:
```python
def add_additional_data(self, instance_data):
    creator_attributes = instance.data.get("creator_attributes", {})
    if creator_attributes.get("render_target") == "farm":
        instance_data["FOV"] = self._context.data["FOV"]

    return instance_data
```

Tilix4 replied:
> Is it changing the logic massively?
I don't understand your question. I want to reuse most of the collector code in a custom plugin, but if I move the whole logic into `plugins/publish` I won't be able to do so because of `__init__.py` limitation. That's why I moved it into the exposed part of the module. To be more specific, I'm inheriting from `CollectRenderBase` to override `get_instances`.

And I got it for the product types and families, I'm waiting we reach an agreement about the design before refactoring it.

iLLiCiTiT replied:
Sorry I'm very busy lately. The issue with the plugin is that it is actually very wrong. It always uses current context instead of instance context (`folder_path`). The collector really should be as simple as possible. It should collect whatever is mandatory to work. The fact it creates instances that do replace existing instances doesn't sound right for me. I know that is because it was like that in OpenPype but it doesn't look good.

I'm not maintainer of Harmony, so I can't tell what is the best solution. From my point of view the plugin logic is wrong as it ignores most of important stuff from reate context, the fact that it does things you need to modify in custom addon sounds like it has more responsibility than it should have. If you want to reuse it by inheriting the plugin then you might hit issues in future releases if we would change anything. DRY is sometimes bad idea, especially if you're working in dynamic addon environment. I don't think the base plugin shoud be an "api" that we should maintain for backwards/forwards compatibility.

BigRoy replied:
> It should collect whatever is mandatory to work. The fact it creates instances that do replace existing instances doesn't sound right for me. I know that is because it was like that in OpenPype but it doesn't look good.

@iLLiCiTiT This is actually an issue best suited for @kalisp with `ayon-deadline` because it's what the base `CollectRender` plug-in unfortunately does. :( The logic is roughly [here](https://github.com/ynput/ayon-core/blob/085c054672098b56cebb91f2007531c077660ef5/client/ayon_core/pipeline/publish/abstract_collect_render.py#L219-L225). It's a bit of a mess really.

iLLiCiTiT replied:
> I understand it from your point of view, this could be solved by adding the possibility to have a __init__.py in the plugins directory, this way I could inherit knowing this is not a stable API

The `__init__.py` would cause it behaves as module which is unwanted. You can do full path import even without `__init__.py` (`import ayon_harmony.plugins.publish.collect_render.CollectRenderBase as CollectRenderBase`).

Tilix4 replied:
Yes I found that very recently, which solves my issue. I made the other required changes, let me know if it's fine for you.

Tilix4 committed:
move product_type resolution from render_target to HarmonyRenderCreator

Tilix4 committed:
restore product type to render

BigRoy commented on client/ayon_harmony/api/plugin.py:
```diff
@@ -186,6 +188,12 @@ class HarmonyRenderCreator(HarmonyCreator):
 
+    def create(self, product_name, instance_data, pre_create_data):
+        # Specify product type from render target
+        self.product_type = f"render.{pre_create_data['render_target']}"
```
> If I oversimplify it: Just use 1 create plugin with render product type and use render_target value to add "render.farm" or "render.local" to "families" during collection phase

This - or do this in the `CreatePlugin` automatically based on change events of the attributes?

Tilix4 committed:
move shared api code to one main plugin

kalisp committed:
Merge branch 'develop' into refactor-collect_render

kalisp reviewed (changes requested):
Render to farm is gone.

Tilix4 committed:
Update client/ayon_harmony/plugins/publish/extract_render.py

lint

Co-authored-by: Jakub Trllo <43494761+iLLiCiTiT@users.noreply.github.com>

Tilix4 committed:
simplify render instance creation remove useless function

Tilix4 committed:
lint

Tilix4 committed:
lint

Tilix4 committed:
Merge branch 'develop' into refactor-collect_render

Tilix4 committed:
fix farm render

Tilix4 commented:
> Render to farm is gone.

Tried a fix, please give it a try

Tilix4 committed:
restore frameStartHandle values from timeline

Tilix4 commented:
> > Render to farm is gone.
> 
> Tried a fix, please give it a try

@kalisp any chance you could try it?

kalisp committed:
Merge remote-tracking branch 'origin/develop' into refactor-collect_render

# Conflicts:
#	client/ayon_harmony/plugins/publish/collect_farm_render.py

kalisp committed:
Fix farm publish

Returned HarmonyRenderInstance with additional fields for farm (Deadline) rendering.

kalisp reviewed (approved):
With https://github.com/ynput/ayon-harmony/pull/50/changes/3263d2c99f8a6e672247ab545c2b39d4ed9169b2 local and farm rendering is working.
